doc/guides/naming_conventions/index.rst

```diff
@@ -0,0 +1,81 @@
+Naming Conventions
+##################
+
+Functions and Procedures
+************************
+
+#. No procedure shall have a name that is a keyword of any standard version of
+   the C or C++ programming language. Restricted names include ``interrupt``, ``inline``,
+   ``class``, ``true``, ``false``, ``public``, ``private``, ``friend``,
+   ``protected``, and many others.
+
+#. Function names are always in lowercase, with an underscore between each part of
+   the name : ``k_timer_init()`` , ``k_timer_start()``.
+
+#. If a function operates primarily on a certain type, the prefix of this function
+   corresponds to the type name. For example, the k_timer_* functions work with the
+   ``k_timer`` type, and k_sem_* functions go with ``k_sem``.
+
+#. Each procedure’s name shall be descriptive of its purpose. Note that procedures
+   encapsulate the “actions” of a program and thus benefit from the use of verbs in
```

Comment:
What I would add here is:
if the module name is the "noun", then "noun-verb" is recommended (eg. `k_mutex_lock`, where `k_mutex` can be considered the module name). However, if the module is not the noun, then the opposite is recommended: `i2c_write()` but `i2c_write_byte()` or `sensor_read()` but `sensor_set_calibration()`.

Reply:
I don't have a strong opinion about this. Doing a quick search in existing .h files in Zephyr verb-noun is used roughly 60% of the time, and noun-verb 40% (very rough estimates). 

One (small) advantage of noun-verb is auto-completion, where nouns are likely more unique than verbs (typically get/set, write/read).

Reply:
I like noun-action since it allows more natural grouping of functions that deal with the same data or concept, with just the last bit being the action. This can extend to more than two levels as well, e.g.

cros_init()
cros_i2c_read()

I also feel that i2c_write_byte() is noun_action since 'write byte' is the action.

Reply:
I like subject-action, recognizing that both subject and action may be complex.  Here's some examples that are causing contention in #29618:
* `k_work_init()` initializes a `k_work` structure
* `k_work_init_delayable()` initializes the delayable variant of a `k_work` structure (i.e. `k_work_delayable`), because `k_work_delayable_init` suggests "perform a potentially delayed initialization of a `k_work` structure", which is not the same thing.

Also what about things that aren't actions, but retrieving state?

* `k_work_busy()` returns `true` iff the work item is being used.  Do we want `k_work_busy_get()`?  I don't....
* `k_work_delayable_remaining()` returns a duration remaining before the delayable `k_work` is submitted.  Appending `_get()` doesn't make it better, IMO.  

There no way any standard is going to eliminate bike shedding, but maybe it will reduce it.

Reply:
this is already covered
> Alternatively, procedures may be named according to the question they answer
   (e.g., ``device_is_ready()``).

so, it should become k_work_is_busy(), k_blah_is_ready() etc...

>     * `k_work_delayable_remaining()` returns a duration remaining before the delayable `k_work` is submitted.  Appending `_get()` doesn't make it better, IMO.

k_work_delayable_remaining_get()

Reply:
OK, I'll leave this be, with a final opinion that using the terminology "subject-action" and "subject-predicate" would be more rigorous and useful.   "noun-verb" wouldn't handle API like:
* `pipeline_put_data()`
* `pipeline_put_control()`